File: _inc/laravel/app/Http/Controllers/Products/ProductServiceCategoryController.php
<?php

namespace App\Http\Controllers\Products;

use App\Config\Constants\{
    DatabaseConstants as DC,
    PermissionsConstants as PMC,
    ViewsConstants as VW,
    ViewClassNamesConstants as VC
};
use App\Http\Controllers\Abstracts\Controller;
use App\Models\{
    Bill,
    ChartOfAccount,
    Invoice,
    ProductService,
    ProductServiceCategory
};
use App\Traits\{ChecksLogin, ChecksPermissions};
use Illuminate\Http\{
    JsonResponse,
    RedirectResponse,
    Request,
    Response
};
use Illuminate\Support\Facades\{
    DB,
    Validator
};
use Illuminate\View\View;
use function App\Http\Controllers\Helpers\{defaultPermissionDenial, defaultUndefinedException};

final class ProductServiceCategoryController extends Controller
{
    use ChecksLogin, ChecksPermissions;

    private const REDIRECT_INDEX = VW::PRD_SV_CAT . '.index';

    public function index(Request $request): View|RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";
        $view = VW::PRD_SV_CAT . '.' . $fn;

        return $this->measureProfile($action, function () use ($request, $view) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, PMC::MNG_CT_CAT, self::REDIRECT_INDEX)) !== true) return $c;
            $categories = ProductServiceCategory::where(DC::COL_TABLE_CREATOR, $user?->creatorId())->get();
            return view($view, compact('categories'));
        });
    }

    public function create(Request $request): View|RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";
        $view = VW::PRD_SV_CAT . '.' . $fn;

        return $this->measureProfile($action, function () use ($request, $view) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, 'create constant category', self::REDIRECT_INDEX)) !== true) return $c;
            $types = ['' => __('Select Category Type')] + ProductServiceCategory::$catTypes;
            $chartAccounts = ChartOfAccount::select(DB::raw('CONCAT(code," - ",name) AS code_name'), 'id')
                ->where(DC::COL_TABLE_CREATOR, $user?->creatorId())
                ->pluck('code_name', 'id')
                ->prepend(__('Select Account'), '');
            return view($view, compact('types', 'chartAccounts'));
        });
    }

    public function store(Request $request): RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";

        return $this->measureProfile($action, function () use ($request) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, 'create constant category', self::REDIRECT_INDEX)) !== true) return $c;
            $validator = Validator::make($request->all(), [
                'name'  => 'required|max:200',
                'type'  => 'required',
                'color' => 'required',
            ]);
            if ($validator->fails()) {
                return redirect()->back()->with('error', $validator->errors()->first());
            }
            $category = new ProductServiceCategory();
            $category->name = $request->input('name');
            $category->type = $request->input('type');
            $category->color = $request->input('color');
            $category->chart_account_id = $request->input('chart_account', 0);
            $category->created_by = $user?->creatorId();
            $category->save();
            return redirect()->route(self::REDIRECT_INDEX)->with('success', __('Category successfully created.'));
        });
    }

    public function edit(Request $request, int|string $id): View|RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";
        $view = VW::PRD_SV_CAT . '.' . $fn;

        return $this->measureProfile($action, function () use ($request, $id, $view) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, 'edit constant category', self::REDIRECT_INDEX)) !== true) return $c;
            $category = ProductServiceCategory::findOrFail($id);
            if ($category->created_by !== $user?->creatorId()) return redirect()->back()->with('error', __('Permission denied.'));
            $types = ProductServiceCategory::$catTypes;
            return view($view, compact('category', 'types'));
        });
    }

    public function update(Request $request, int|string $id): RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";

        return $this->measureProfile($action, function () use ($request, $id) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, 'edit constant category', self::REDIRECT_INDEX)) !== true) return $c;
            $category = ProductServiceCategory::findOrFail($id);
            if ($category->created_by !== $user?->creatorId()) return redirect()->back()->with('error', __('Permission denied.'));
            $validator = Validator::make($request->all(), [
                'name'  => 'required|max:200',
                'type'  => 'required',
                'color' => 'required',
            ]);
            if ($validator->fails()) {
                return redirect()->back()->with('error', $validator->errors()->first());
            }
            $category->fill([
                'name'             => $request->input('name'),
                'type'             => $request->input('type'),
                'color'            => $request->input('color'),
                'chart_account_id' => $request->input('chart_account', 0),
            ])->save();
            return redirect()->route(self::REDIRECT_INDEX)->with('success', __('Category successfully updated.'));
        });
    }

    public function destroy(Request $request, int|string $id): RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";

        return $this->measureProfile($action, function () use ($request, $id) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, 'delete constant category', self::REDIRECT_INDEX)) !== true) return $c;
            $category = ProductServiceCategory::findOrFail($id);
            if ($category->created_by !== $user?->creatorId()) return redirect()->back()->with('error', __('Permission denied.'));
            $existsCheck = match ($category->type) {
                0 => ProductService::where('category_id', $id)->exists(),
                1 => Invoice::where('category_id', $id)->exists(),
                default => Bill::where('category_id', $id)->exists(),
            };
            if ($existsCheck) return redirect()->back()->with('error', __('This category is in use.'));
            $category->delete();
            return redirect()->route(self::REDIRECT_INDEX)->with('success', __('Category successfully deleted.'));
        });
    }

    public const GET_PRD_CAT = 'getProductCategories';
    public function getProductCategories(Request $request): Response|RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";

        return $this->measureProfile($action, function () use ($request) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, PMC::MNG_CT_CAT, self::REDIRECT_INDEX)) !== true) return $c;

            $html = '<div class="mb-3 mr-2 zoom-in ">
        <div class="' . VC::CD . ' rounded-10 card-stats mb-0 cat-active overflow-hidden" data-id="0">
            <div class="category-select" data-cat-id="0">
                    <button type="button" class="btn tab-btns btn-primary">'
                . __("All Categories") .
                '</button>
                </div>
            </div>
        </div>';

            foreach (ProductServiceCategory::where(DC::COL_TABLE_CREATOR, $user?->creatorId())->get() as $c) {
                $html .= '<div class="mb-3 mr-2 zoom-in cat-list-btn">
            <div class="' . VC::CD . ' rounded-10 card-stats mb-0 overflow-hidden" data-id="'
                    . $c->id . '">
            <div class="category-select" data-cat-id="'
                    . $c->id . '">
                <button type="button" class="btn tab-btns btn-primary">'
                    . $c->name .
                    '</button>
            </div>
            </div>
        </div>';
            }

            return response($html);
        });
    }

    public const GET_ACC = 'getAccount';
    public const IDX = 'index';
    public const CRT = 'create';
    public const STR = 'store';
    public const EDT = 'edit';
    public const UPD = 'update';
    public const DEL = 'destroy';

    public function getAccount(Request $request): JsonResponse|RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";

        return $this->measureProfile($action, function () use ($request) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, PMC::MNG_CT_CAT, self::REDIRECT_INDEX)) !== true) return $c;

            $map = [
                'income'             => 'Income',
                'expense'            => 'Expenses',
                'asset'              => 'Assets',
                'liability'          => 'Liabilities',
                'equity'             => 'Equity',
                'costs of good sold' => 'Costs of Goods Sold',
            ];
            $typeName = $map[$request->type] ?? null;

            $chartAccounts = $typeName
                ? ChartOfAccount::select(DB::raw('CONCAT(code," - ",name) AS code_name'), 'id')
                ->leftJoin('chart_of_account_types', 'chart_of_account_types.id', 'chart_of_accounts.type')
                ->where('chart_of_account_types.name', $typeName)
                ->where('chart_of_accounts.created_by', $user?->creatorId())
                ->pluck('code_name', 'id')
                : [];

            return response()->json($chartAccounts);
        });
    }

    /**
     * Show a single product/service category.
     */
    public function show(Request $request, int|string $id): View|RedirectResponse|JsonResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";
        $view = VW::PRD_SV_CAT . '.' . $fn;
        return $this->measureProfile($action, function () use ($request, $id, $view, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, PMC::MNG_CT_CAT, self::REDIRECT_INDEX)) !== true) return $c;
            try {
                $category = ProductServiceCategory::where(DC::COL_TABLE_CREATOR, $user?->creatorId())->findOrFail($id);
                if ($request->wantsJson()) return response()->json($category);
                if (!\Illuminate\Support\Facades\View::exists($view)) {
                    return redirect()->route(self::REDIRECT_INDEX)->with('info', __('Category detail view not available.'));
                }
                return view($view, compact('category'));
            } catch (\Illuminate\Database\Eloquent\ModelNotFoundException $e) {
                return redirect()->route(self::REDIRECT_INDEX)->with('error', __('Category not found.'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($request, $e, "$action");
            }
        });
    }
}

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/products/product_service_category_controller.py
import inspect
import logging
from django.contrib import messages
from django.db.models import Q
from django.http import HttpRequest, JsonResponse
from django.shortcuts import redirect, render
from django.utils.decorators import method_decorator
from django.views.decorators.http import require_GET, require_POST
from ....Models.bills.bill import Bill
from ....Models.bills.invoice import Invoice
from ....Models.charts.chart_of_account import ChartOfAccount
from ....Models.products.product import Product
from ....Models.products.product_service import ProductService
from ....Models.products.product_service_category import ProductServiceCategory
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from .._helpers.http import get_redirect_url
from .._helpers.security import permission_required_custom
from django.contrib.auth.decorators import login_required
from .._traits.controller import Controller

logger = logging.getLogger(__name__)

class ProductServiceCategoryController(Controller):

  @classmethod
  @method_decorator(login_required)
  @permission_required_custom('manage constant category',logger)
  @require_GET
  def index(cls,request:HttpRequest):
    try:
      cats=ProductServiceCategory.objects.filter(
        created_by=request.user.creator_id()
      )
      return render(request,
                    'productServiceCategory/index.html',
                    {'categories':cats})
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

  @classmethod
  @method_decorator(login_required)
  @permission_required_custom('create constant category',logger)
  @require_GET
  def create(cls,request:HttpRequest):
    try:
      types_opts={'':'Select Category Type',**ProductServiceCategory.catTypes}
      chart_accounts=ChartOfAccount.objects.filter(
        created_by=request.user.creator_id()
      ).values_list('id','code')
      return render(request,
                    'productServiceCategory/create.html',
                    {'types':types_opts,
                     'chart_accounts':chart_accounts})
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

  @classmethod
  @method_decorator(login_required)
  @permission_required_custom('create constant category',logger)
  @require_POST
  def store(cls,request:HttpRequest):
    try:
      d=request.POST
      name=d.get('name','').strip()
      color=d.get('color','').strip()
      typeVal=d.get('type','').strip()
      if not name:
        messages.error(request,"Name is required");
        return redirect(get_redirect_url(request))
      if len(name)>200:
        messages.error(request,"Name must be at most 200 characters");
        return redirect(get_redirect_url(request))
      if not color:
        messages.error(request,"Color is required");
        return redirect(get_redirect_url(request))
      if len(color)>100:
        messages.error(request,"Color must be at most 100 characters");
        return redirect(get_redirect_url(request))
      if not typeVal:
        messages.error(request,"Type is required");
        return redirect(get_redirect_url(request))
      if typeVal not in ProductServiceCategory.catTypes:
        messages.error(request,"Invalid category type");
        return redirect(get_redirect_url(request))
      cat=ProductServiceCategory()
      cat.name=name
      cat.color=color
      cat.type=typeVal
      cat.chart_account_id=d.get('chart_account') or '0'
      cat.created_by=request.user.creator_id()
      cat.save()
      return redirect(get_redirect_url(request))
    except Exception as e:
      return default_permission_denial(
        request,
        err=e,
        ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

  @classmethod
  @method_decorator(login_required)
  @permission_required_custom('view constant category',logger)
  @require_GET
  def show(cls,request:HttpRequest,pk):
    try:
      cat=ProductServiceCategory.objects.get(
        pk=pk,created_by=request.user.creator_id()
      )
      return render(request,
                    'productServiceCategory/show.html',
                    {'category':cat})
    except ProductServiceCategory.DoesNotExist as e:
      return default_permission_denial(
        request,
        err=e,
        ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

  @classmethod
  @method_decorator(login_required)
  @permission_required_custom('edit constant category',logger)
  @require_GET
  def edit(cls,request:HttpRequest,pk):
    try:
      cat=ProductServiceCategory.objects.get(pk=pk)
      return render(request,
                    'productServiceCategory/edit.html',
                    {'category':cat,'types':ProductServiceCategory.catTypes})
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

  @classmethod
  @method_decorator(login_required)
  @permission_required_custom('edit constant category',logger)
  @require_POST
  def update(cls,request:HttpRequest,pk):
    try:
      d=request.POST
      cat=ProductServiceCategory.objects.get(
        pk=pk,created_by=request.user.creator_id()
      )
      name=d.get('name','').strip()
      color=d.get('color','').strip()
      typeVal=d.get('type','').strip()
      if not name:
        messages.error(request,"Name is required");
        return redirect(get_redirect_url(request))
      if len(name)>200:
        messages.error(request,"Name must be at most 200 characters");
        return redirect(get_redirect_url(request))
      if not color:
        messages.error(request,"Color is required");
        return redirect(get_redirect_url(request))
      if len(color)>100:
        messages.error(request,"Color must be at most 100 characters");
        return redirect(get_redirect_url(request))
      if not typeVal:
        messages.error(request,"Type is required");
        return redirect(get_redirect_url(request))
      if typeVal not in ProductServiceCategory.catTypes:
        messages.error(request,"Invalid category type");
        return redirect(get_redirect_url(request))
      cat.name=name
      cat.color=color
      cat.type=typeVal
      cat.chart_account_id=d.get('chart_account') or 0
      cat.save()
      return redirect(get_redirect_url(request))
    except Exception as e:
      return default_permission_denial(
        request,
        err=e,
        ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

  @classmethod
  @method_decorator(login_required)
  @permission_required_custom('delete constant category',logger)
  @require_POST
  def destroy(cls,request:HttpRequest,pk):
    try:
      cat=ProductServiceCategory.objects.get(
        pk=pk,created_by=request.user.creator_id()
      )
      rel=(Bill.objects.filter(category_id=pk).first()
           if cat.type=='bill'
           else Invoice.objects.filter(category_id=pk).first()
           if cat.type=='invoice'
           else Product.objects.filter(category_id=pk).first())
      if rel:
        return redirect(get_redirect_url(request))
      cat.delete()
      return redirect(get_redirect_url(request))
    except Exception as e:
      return default_permission_denial(
        request,
        err=e,
        ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

  @classmethod
  @method_decorator(login_required)
  @require_GET
  def get_product_categories(cls,request:HttpRequest):
    try:
      html='<div class="mb-3 mr-2 zoom-in">'\
           '<div class="card rounded-10 card-stats mb-0 cat-active overflow-hidden" data-id="0">'\
           '<div class="category-select" data-cat-id="0">'\
           '<button type="button" class="btn tab-btns btn-primary">All Categories</button>'\
           '</div></div></div>'
      for c in ProductServiceCategory.get_all_categories(request.user):
        html+=f'<div class="mb-3 mr-2 zoom-in cat-list-btn">'\
              f'<div class="card rounded-10 card-stats mb-0 overflow-hidden" data-id="{c.id}">'\
              f'<div class="category-select" data-cat-id="{c.id}">'\
              f'<button type="button" class="btn tab-btns btn-primary">{c.name}</button>'\
              '</div></div></div>'
      return JsonResponse({'html':html})
    except Exception as e:
      logger.error(f'{cls.__name__}::get_product_categories failed: {e}')
      return JsonResponse({'html':''},status=500)

  @classmethod
  @method_decorator(login_required)
  @require_GET
  def search_products_by_name(cls,request:HttpRequest):
    try:
      term=request.GET.get('q','').strip()
      ps=ProductService.objects.filter(
        name__icontains=term,
        created_by=request.user.creator_id()
      ).values('id','name')[:10]
      pr=Product.objects.filter(
        name__icontains=term,
        created_by=request.user.creator_id()
      ).values('id','name')[:10]
      res=[{'id':x['id'],'name':x['name'],'type':'service'}for x in ps]
      res+=[{'id':x['id'],'name':x['name'],'type':'product'}for x in pr]
      return JsonResponse(res,safe=False)
    except Exception as e:
      logger.error(f'{cls.__name__}::search_products_by_name failed: {e}')
      return JsonResponse([],status=500)

  @classmethod
  @method_decorator(login_required)
  @require_POST
  def get_account(cls,request:HttpRequest):
    try:
      t=request.POST.get('type')
      q=ChartOfAccount.objects.filter(
        created_by=request.user.creator_id()
      )
      mapping={
        'income':Q(type__name='Income'),
        'expense':Q(type__name='Expenses'),
        'asset':Q(type__name='Assets'),
        'liability':Q(type__name='Liabilities'),
        'equity':Q(type__name='Equity'),
        'costs of good sold':Q(type__name='Costs of Goods Sold')
      }
      acc=list(q.filter(mapping.get(t,Q())).values('id','code'))
      return JsonResponse(acc,safe=False)
    except Exception as e:
      logger.error(f'{cls.__name__}::get_account failed: {e}')
      return JsonResponse([],status=500)
